### PR TITLE
Adjust or center startup lists in home buffer

### PR DESCRIPTION
**Pull request by deb0ch**
Just centering startup lists would look weird, especially when all the entries are short: the beginning would be almost in the middle of the page.

Moreover, an inconsistent position of the lists would make them less easy to read.

This is why the position of the beginning of the startup lists is relative to the position of the home buffer buttons, unless the lines are too long to fit the screen. In that case, the lists are centered.

### CODE REVIEW

**Comment by deb0ch** [comment]
Damn, will take care of the conflicts very soon.

**Comment by deb0ch** [comment]
By the way, I used `(window-width)` instead of `spacemacs-buffer--banner-length` that is usually used for centering because I don't really understand the point of it, and it looks like being just the same value.

**Comment by syl20bnr** [comment]
Can you post before/after screenshots ?

**Comment by deb0ch** [comment]
Hey, sure !

Before:
screenshot from 2016-09-04 21-35-34

After:
screenshot from 2016-09-04 21-36-58

screenshot from 2016-09-04 21-43-17

screenshot from 2016-09-04 21-46-52

When it wouldn't fit the screen anymore, it gets centered:

screenshot from 2016-09-04 23-08-08

**Comment by JAremko** [comment]
@deb0ch  looks centered to me :+1:

File: core/core-spacemacs-buffer.el [hunk 1]
```diff
@@ -590,7 +590,11 @@ border."
                    :mouse-face 'highlight
                    :follow-link "\C-m"
                    (propertize "Update Spacemacs" 'face 'font-lock-keyword-face))
-    (spacemacs-buffer//center-line)
+    (let ((len (- (line-end-position) (line-beginning-position))))
+      (spacemacs-buffer//center-line)
+      (setq spacemacs-buffer--buttons-position (- (- (line-end-position)
```
Comment (TheBB) [inline core/core-spacemacs-buffer.el]: `-` accepts more than one argument.

File: core/core-spacemacs-buffer.el [hunk 2]
```diff
@@ -828,9 +826,53 @@ list. Return entire list if `END' is omitted."
...
+(defun spacemacs-buffer//center-startupify-lists ()
+  (let* ((lists-width (spacemacs-buffer//get-buffer-width))
+         (margin (max 0 (- spacemacs-buffer--buttons-position
+                           spacemacs-buffer-buttons-recent-lists-offset)))
+         (final-padding (if (< spacemacs-buffer--banner-length (+ margin lists-width))
+                            (max 0 (floor (/ (- spacemacs-buffer--banner-length lists-width) 2)))
+                          margin)))
+    (goto-char 0)
+    (while (not (eobp))
+      (progn (line-beginning-position)
```
Comment (TheBB) [inline core/core-spacemacs-buffer.el]: No need for `progn`.

File: core/core-spacemacs-buffer.el [hunk 3]
```diff
@@ -828,9 +826,53 @@ list. Return entire list if `END' is omitted."
...
+          (if (< current-max line-length)
+              (setq current-max line-length)))
+        (forward-line 1))
+      current-max)))
+
+(defvar spacemacs-buffer--buttons-position nil
+  "Offset in characters between the edge of the screen and the beginning of the
+ home buffer buttons. Do not try to set this variable.")
+
+(defvar spacemacs-buffer-buttons-recent-lists-offset 25
```
Comment (TheBB) [inline core/core-spacemacs-buffer.el]: This seems to be independent of the actual width of the startup lists, i.e. not really centered. Am I missing something?

In any case this should probably not have `recent` in the name.

Would prefer to move the defvars to the top of the file if possible.

**Comment by TheBB** [comment]
Please do excuse me for taking my time with this, but last time we went on a centering spree we dragged a ton of bugs in, for very little (IMO) gain, at a time when we're waiting for release. I'm not too keen on going through that again.

File: core/core-spacemacs-buffer.el (hunk 1 above)
Comment (deb0ch) [inline core/core-spacemacs-buffer.el]: Changed, thanks for the tip !

File: core/core-spacemacs-buffer.el (hunk 3 above)
Comment (deb0ch) [inline core/core-spacemacs-buffer.el]: Yes, it is independent of the width of the startup lists.  
The reason is that just centering based on the width of the startup lists would feel weird, so I did not just center it.  

The positioning algorithm is as follows:
- if there is enough space, the startup lists will have a fixed relative position to the actually centered elements (here the home buffer buttons)
- if a part of the startup lists would be off the screen just center them regularly, based on the width of the startup lists and the width of the window.

This way, the position of the beginning of the lines is consistent which makes reading easier, and the width of the buffer is always used at its best even when some entries are very long.

Comment (deb0ch) [inline core/core-spacemacs-buffer.el]: So, sometimes it is centered, sometimes it is adjusted.

Comment (TheBB) [inline core/core-spacemacs-buffer.el]: Yeah that's a fair point.

**Comment by TheBB** [comment]
Thanks! Cherry-picked in develop. You can safely delete your branch.

**Pull request closed without merging**